### ISSUE TITLE
Consider an alternative to `random_32_bytes`

### ISSUE THREAD

**Issue by Kixunil** [rust-bitcoin/rust-secp256k1#522]
I don't like non-generic `32_bytes` much. I was thinking of some alternative designs: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=3123fb72993eabdeeb4a51eb2558be88

_Originally posted by @Kixunil in https://github.com/rust-bitcoin/rust-secp256k1/pull/520#discussion_r1026403543_

**Comment by apoelstra** [comment]
There are several places in this library where we need 32 uniformly random bytes, and zero where we need a different number of bytes or a different distribution. In fact if we ever use a different distribution this is nearly guaranteed to be a key-leaking catastrophe. If we generate too few bytes this is also a catastrophe. If we generate too many I guess that's fine, though depending how we implement things it could cause panics or something.

Secondly, the `Standard` distribution is documented as "usually" being uniform. We should use `Uniform` if we mean uniform.

Thirdly, I worry that using all these generics and stuff will make it hard to move away from `rand` in the future if we eventually find a simpler crate that can wrap `getrandom` and be more stable.

Basically, I think any move away from "always generate 32 bytes uniformly" will create an ongoing maintenance/review burden as we manually check that the new generic code is only ever used to uniformly generate 32 bytes.

**Comment by Kixunil** [comment]
> Secondly, the `Standard` distribution is documented as "usually" being uniform. We should use `Uniform` if we mean uniform.

Oh, interesting footgun, I totally missed that. Yeah, it'd have to be `Uniform`.

> if we eventually find a simpler crate that can wrap `getrandom` and be more stable.

Unless that crate is `rand_core` this could be detrimental due to network effects.

Another approach is to only call `T::random()` inherent method from all code and have `fill_bytes` in its impl - just like `Scalar` does. So there should be exactly three calls to `random_32_bytes`/`fill_bytes` in the whole codebase.

**Comment by apoelstra** [comment]
> Another approach is to only call T::random() inherent method from all code and have fill_bytes in its impl - just like Scalar does. 

Yeah, I wouldn't mind this.
